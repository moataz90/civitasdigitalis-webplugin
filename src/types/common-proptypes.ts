export interface IPropsChildable {
	children?: any;
}

export interface IPropsStyledComponents {
	className?: string;
}