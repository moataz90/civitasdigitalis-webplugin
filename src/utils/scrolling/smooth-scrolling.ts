import { EasingFunction } from './easing';

export type ScrollSmothCallback = (interpolatedValue: number) => void;

export const scrollSmooth = (startValue: number, targetValue: number, interpolation: EasingFunction, animationTime: number,
	callbackUpdate: ScrollSmothCallback, callbackFinish?: () => void) => {

	const diff = targetValue - startValue;

	let lastTime = Date.now();
	let timeSum = 0;

	const internalInterpolate = () => {
		const p = Math.min(timeSum / animationTime, 1);
		const newValue = diff * interpolation(p);

		callbackUpdate(diff > 0 ? Math.min(startValue + newValue, targetValue) : Math.max(startValue + newValue, targetValue));

		timeSum += (Date.now() - lastTime);
		lastTime = Date.now();

		if (p < 1) {
			window.requestAnimationFrame(internalInterpolate);
		} else {
			if (callbackFinish) {
				callbackFinish();
			}
		}
	};

	window.requestAnimationFrame(internalInterpolate);
};