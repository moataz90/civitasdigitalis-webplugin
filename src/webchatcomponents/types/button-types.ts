export enum MessageButtonType {
	Url,
	Postback,
	Call,
	Share
}

export interface IMessagePayloadButton {
	type: MessageButtonType;
	title: string;
	payload?: string;
}

export type ButtonCallback = (type: number, title: string, payload: any) => void;