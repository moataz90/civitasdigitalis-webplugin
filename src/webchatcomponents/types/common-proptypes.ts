import * as React from 'react';

export interface IPropsChildable {
	children?: any;
}

export interface IPropsStyledComponents {
	className?: string;
}

export interface IPropsCustomStyleable {
	css?: React.CSSProperties;
}