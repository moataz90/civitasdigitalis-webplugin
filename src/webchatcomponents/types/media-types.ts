export enum MessageMediaType {
	Audio,
	Video,
	Image,
	File
}