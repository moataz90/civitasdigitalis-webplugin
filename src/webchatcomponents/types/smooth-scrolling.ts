export type EasingFunction = (t: number) => number;

export type ScrollSmothCallback = (interpolatedValue: number) => void;

export type SmoothScrolling = (
	startValue: number,
	targetValue: number,
	interpolation: EasingFunction,
	animationTime: number,
	callbackUpdate: ScrollSmothCallback,
	callbackFinish?: () => void
) => void;