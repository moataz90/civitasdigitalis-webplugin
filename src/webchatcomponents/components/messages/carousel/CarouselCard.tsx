import * as React from 'react';
/* tslint:disable:no-unused-variable */
import styled, { StyledComponentClass } from 'styled-components';
import { IPropsStyledComponents } from '../../../types/common-proptypes';
import { CarouselCardImage } from './CarouselCardImage';
import { CarouselCardContent } from './CarouselCardContent';
import { CarouselCardTitle } from './CarouselCardTitle';
import { CarouselCardSubtitle } from './CarouselCardSubtitle';
import { detect, BrowserInfo } from 'detect-browser';

const browser: BrowserInfo =  {
	name: 'ios',
	version: '1.0',
	os: 'windows'
};

/* istanbul ignore next */
const detectBrowser: BrowserInfo = detect() === null ? browser : detect() ;
/* istanbul ignore next */
const roundBorders = (detectBrowser.name === 'ios') ? true : false;

const CarouselCardWrapper = styled.div`
	&&& {
		margin-left: 10px;
		display: flex;
		flex-direction: column;
		flex-shrink: 0;
		padding-bottom: 8px;
	}

	&&&:first-child{
		margin-left: 0px;
	}

	&&&:last-child{
		border-right: 10px solid transparent;
	}
`;

export interface ICarouselCardProps extends IPropsStyledComponents {
	children: React.ReactNode[];
	title: string;
	subtitle: string;
	imgUrl: string;
	fontFamily: string;
	fontColorTitle: string;
	fontColorSubtitle: string;
	isFirst?: boolean;
	isLast?: boolean;
}

const CarouselCardComponent: React.StatelessComponent<ICarouselCardProps> = (props) => (
	<CarouselCardWrapper>
		<div className={props.className}>
			<CarouselCardImage imageUrl={props.imgUrl} />
			<CarouselCardContent>
				<CarouselCardTitle title={props.title} fontFamily={props.fontFamily} fontColor={props.fontColorTitle} />
				<CarouselCardSubtitle subtitle={props.subtitle} fontFamily={props.fontFamily} fontColor={props.fontColorSubtitle} />
			</CarouselCardContent>
		</div>
		{props.children}
	</CarouselCardWrapper>
);
/* istanbul ignore next */
const CarouselCardStyled = styled(CarouselCardComponent)`
	&&& {
		width: 260px;
	
		border-top-left-radius: 16px;
		border-top-right-radius: 16px;
		border-bottom-left-radius: ${props => (props.isFirst) ? roundBorders ? '16px' : '0px' : '16px'};
		border-bottom-right-radius: ${props => (props.isLast) ? roundBorders ? '16px' : '0px' : '16px'};
		background-color: white;
		box-shadow: 0px 0px 10px -2px rgba(0,0,0,0.3);
		margin: 15px 10px;
		margin-right: 0px;
		overflow: hidden;
	}
`;

export const CarouselCard = CarouselCardStyled;