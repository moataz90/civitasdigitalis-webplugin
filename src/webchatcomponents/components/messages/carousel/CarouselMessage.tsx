import * as React from 'react';
/* tslint:disable:no-unused-variable */
import styled, { StyledComponentClass } from 'styled-components';
import { Message, IMessageBaseProps } from '../Message';
import * as ReactDOM from 'react-dom';
import { CarouselArrowLeft, CarouselArrowRight } from './CarouselArrow';
import { SmoothScrolling, EasingFunction } from '../../../types/smooth-scrolling';

const ScrollWrapperOuter = styled.div`
	&&& {
		width: 100%;
		overflow: hidden;
	}
`;

const ScrollWrapperInner = styled.div`
	&&& {
		height: 100%;
		position: relative;
		padding-bottom: 0px;
		box-sizing: border-box;
		height: calc(100% + 17px);
    	margin-bottom: -17px; /* -17 for IE11*/
	}
`;

const CardWrapper = styled.div`
	&&& {
		overflow-y: hidden;
		overflow-x: scroll;
		display: flex;
		flex-direction: row;
		padding-bottom: 20px;
	}
`;

export interface ICarouselMessageProps extends IMessageBaseProps {
	children: React.ReactNode[];
	smoothScrollingFunction: SmoothScrolling;
	easing: EasingFunction;
	canScrollLeft?: boolean;
	canScrollRight?: boolean;
}

export interface ICarouselMessageState {
	scrollTarget: number;
	isAnimating: boolean;
	canScrollLeft: boolean;
	canScrollRight: boolean;
}

const CAROUSEL_CARD_WIDTH = 280;

export class CarouselMessageComponent extends React.Component<ICarouselMessageProps, ICarouselMessageState> {
	private cardWrapper: HTMLDivElement;
	private oldScrollLeft: number = 0;

	constructor(props: ICarouselMessageProps) {
		super(props);

		this.state = {
			...this.state,
			scrollTarget: 0,
			isAnimating: false,
			canScrollLeft: props.canScrollLeft || false,
			canScrollRight: props.canScrollRight || false
		};

		this.onClickArrow = this.onClickArrow.bind(this);
		this.onScrollCardWrapper = this.onScrollCardWrapper.bind(this);
	}

	/* istanbul ignore next */
	public componentDidMount() {
		if (this.cardWrapper) {
			this.checkArrowVisibility();

			if (this.cardWrapper.addEventListener && window.addEventListener) {
				this.cardWrapper.addEventListener('scroll', this.onScrollCardWrapper);
				window.addEventListener('resize', this.onScrollCardWrapper);
			}
		}
	}

	/* istanbul ignore next */
	public componentWillUnmount() {
		this.cardWrapper.removeEventListener('scroll', this.onScrollCardWrapper);
		window.removeEventListener('resize', this.onScrollCardWrapper);
	}

	public render() {
		const { children, isLastMessage } = this.props;
		const { canScrollLeft, canScrollRight } = this.state;

		return (
			<Message isOwnMessage={false} css={{ padding: 0, paddingBottom: 10 }} isLastMessage={isLastMessage}>
				<ScrollWrapperOuter>
					<ScrollWrapperInner>
						{canScrollLeft
							? <CarouselArrowLeft isLeftArrow={true} onClick={this.onClickArrow} />
							: null
						}
						{canScrollRight
							? <CarouselArrowRight isLeftArrow={false} onClick={this.onClickArrow} />
							: null
						}
						<CardWrapper
							ref={(instance: React.ReactInstance) =>
								this.cardWrapper = ReactDOM.findDOMNode(instance) as HTMLDivElement}
						>
							{children}
						</CardWrapper>
					</ScrollWrapperInner>
				</ScrollWrapperOuter>
			</Message>
		);
	}

	/* istanbul ignore next */
	private onScrollCardWrapper(event: UIEvent) {
		this.checkArrowVisibility();
	}

	/* istanbul ignore next */
	private checkArrowVisibility() {
		if (this.state.isAnimating) {
			return;
		}

		const { children } = this.props;

		const currentScrollLeftPosition = this.cardWrapper.scrollLeft;
		const maxScrollLeftPosition = children.length * CAROUSEL_CARD_WIDTH;

		const canScrollLeft = currentScrollLeftPosition > 0;
		const canScrollRight = children.length > 1 && currentScrollLeftPosition <
			this.cardWrapper.scrollWidth - this.cardWrapper.clientWidth;

		let newCanScrollLeft = this.state.canScrollLeft;
		let newCanScrollRight = this.state.canScrollRight;

		if (canScrollLeft !== this.state.canScrollLeft) {
			newCanScrollLeft = canScrollLeft;
		}

		if (canScrollRight !== this.state.canScrollRight) {
			newCanScrollRight = canScrollRight;
		}

		this.setState({
			...this.state,
			canScrollLeft: newCanScrollLeft,
			canScrollRight: newCanScrollRight,
			scrollTarget: this.cardWrapper.scrollLeft
		});
	}

	/* istanbul ignore next */
	private onClickArrow(arrowLeft: boolean) {
		if (this.state.isAnimating) {
			return;
		}

		let updateState = false;
		const currentScrollIdx = Math.floor(this.cardWrapper.scrollLeft / CAROUSEL_CARD_WIDTH);
		const { children, smoothScrollingFunction, easing } = this.props;

		let newCardScrollIdx = currentScrollIdx;

		if (arrowLeft && currentScrollIdx > 0) {
			// Scroll left is possible
			if (this.cardWrapper.scrollLeft % CAROUSEL_CARD_WIDTH === 0) {
				newCardScrollIdx--;
			}

			updateState = true;
		} else if (!arrowLeft && currentScrollIdx < children.length - 1) {
			// Scroll right is possible
			newCardScrollIdx++;

			updateState = true;
		}

		const newScrollTarget = Math.min(
			newCardScrollIdx * CAROUSEL_CARD_WIDTH,
			this.cardWrapper.scrollWidth - this.cardWrapper.clientWidth
		);

		if (this.cardWrapper.scrollLeft !== newScrollTarget) {
			this.setState({
				...this.state,
				isAnimating: true,
			});

			const scrollingCallback = () => {
				this.setState({
					...this.state,
					isAnimating: false,
					scrollTarget: newScrollTarget
				},
					() => {
						this.checkArrowVisibility();
					});
			};

			smoothScrollingFunction(this.cardWrapper.scrollLeft, newScrollTarget, easing, 500, (newVal) => {
				this.cardWrapper.scrollLeft = newVal;
			}, scrollingCallback);
		}
	}
}

export const CarouselMessage = CarouselMessageComponent;