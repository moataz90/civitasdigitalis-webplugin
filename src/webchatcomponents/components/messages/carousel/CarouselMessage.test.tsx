import * as React from 'react';
import * as renderer from 'react-test-renderer';
import 'jest-styled-components';
import { CarouselMessage } from './CarouselMessage';
import { EasingFunction } from '../../../types/smooth-scrolling';
import { CarouselCard } from './CarouselCard';

// The following test cases are based on jest snapshot testing!
// For more information on how snapshot tests are working, see
// https://jest-bot.github.io/jest/docs/snapshot-testing.html

const smoothScrollingFuncMock = (startValue: number, targetValue: number, interpolation: EasingFunction) => {
	// Do nothing
};

test('render carousel message', () => {
	const cardData = ['Card1', 'Card2', 'Card3'];

	const domTree = renderer.create(
		<CarouselMessage
			smoothScrollingFunction={smoothScrollingFuncMock}
			easing={(t): number => { return 0; }}
			isOwnMessage={true}
			isLastMessage={false}
		>
			{
				cardData.map((cardInfo, idx) => (
					<CarouselCard
						title={cardInfo}
						subtitle={cardInfo}
						imgUrl="someurl"
						fontFamily="Helvetica"
						fontColorTitle="black"
						fontColorSubtitle="white"
						isFirst={idx === 0}
						isLast={idx === cardData.length - 1}
						key={idx}
					>
						<div>Some child</div>
						<div>Another child</div>
					</CarouselCard>

				))
			}
		</CarouselMessage>
	).toJSON();

	expect(domTree).toMatchSnapshot();
});

test('render carousel message canScrollLeft', () => {
	const cardData = ['Card1', 'Card2', 'Card3'];

	const domTree = renderer.create(
		<CarouselMessage
			smoothScrollingFunction={smoothScrollingFuncMock}
			easing={(t): number => { return 0; }}
			isOwnMessage={true}
			isLastMessage={false}
			canScrollLeft={true}
			canScrollRight={true}
		>
			{
				cardData.map((cardInfo, idx) => (
					<CarouselCard
						title={cardInfo}
						subtitle={cardInfo}
						imgUrl="someurl"
						fontFamily="Helvetica"
						fontColorTitle="black"
						fontColorSubtitle="white"
						isFirst={idx === 0}
						isLast={idx === cardData.length - 1}
						key={idx}
					>
						<div>Some child</div>
						<div>Another child</div>
					</CarouselCard>

				))
			}
		</CarouselMessage>
	).toJSON();

	expect(domTree).toMatchSnapshot();
});