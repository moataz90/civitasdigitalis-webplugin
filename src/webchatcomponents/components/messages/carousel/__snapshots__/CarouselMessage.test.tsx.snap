// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render carousel message 1`] = `
.c0.c0.c0 {
  width: 100%;
  position: relative;
  box-sizing: border-box;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: flex-end;
  -webkit-box-align: flex-end;
  -ms-flex-align: flex-end;
  align-items: flex-end;
  -webkit-flex-flow: column nowrap;
  -ms-flex-flow: column nowrap;
  flex-flow: column nowrap;
  -webkit-box-pack: end;
  -webkit-justify-content: flex-end;
  -ms-flex-pack: end;
  justify-content: flex-end;
  padding-right: 10px;
}

.c1.c1.c1 {
  width: 100%;
  overflow: hidden;
}

.c2.c2.c2 {
  height: 100%;
  position: relative;
  padding-bottom: 0px;
  box-sizing: border-box;
  height: calc(100% + 17px);
  margin-bottom: -17px;
}

.c3.c3.c3 {
  overflow-y: hidden;
  overflow-x: scroll;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  padding-bottom: 20px;
}

.c6.c6.c6 {
  width: 100%;
  border-bottom: 0px;
  box-sizing: border-box;
  overflow: hidden;
  position: relative;
}

.c6.c6.c6:after {
  padding-top: 52%;
  display: block;
  content: '';
}

.c7.c7.c7 {
  position: absolute;
  top: 0;
  bottom: 0;
  right: 0;
  left: 0;
  background-repeat: no-repeat;
  background-position: 50% 50%;
  background-size: cover;
  background-image: url(someurl);
}

.c8.c8.c8 {
  width: 100%;
  padding: 8px 6px;
  padding-bottom: 12px;
  box-sizing: border-box;
  height: 92px;
  line-height: normal;
}

.c9.c9.c9 {
  font-size: 14px;
  font-family: Helvetica;
  font-weight: bold;
  overflow: hidden;
  color: black;
}

.c10.c10.c10 {
  font-size: 12px;
  font-family: Helvetica;
  color: white;
}

.c4.c4.c4 {
  margin-left: 10px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  padding-bottom: 8px;
}

.c4.c4.c4:first-child {
  margin-left: 0px;
}

.c4.c4.c4:last-child {
  border-right: 10px solid transparent;
}

.c5.c5.c5 {
  width: 260px;
  border-top-left-radius: 16px;
  border-top-right-radius: 16px;
  border-bottom-left-radius: 16px;
  border-bottom-right-radius: 16px;
  background-color: white;
  box-shadow: 0px 0px 10px -2px rgba(0,0,0,0.3);
  margin: 15px 10px;
  margin-right: 0px;
  overflow: hidden;
}

<div
  className="c0"
  style={
    Object {
      "padding": 0,
      "paddingBottom": 10,
    }
  }
>
  <div
    className="c1"
  >
    <div
      className="c2"
    >
      <div
        className="c3"
      >
        <div
          className="c4"
        >
          <div
            className="c5"
          >
            <div
              className="c6"
            >
              <div
                className="c7"
              />
            </div>
            <div
              className="c8"
            >
              <div
                className="c9"
              >
                Card1
              </div>
              <div
                className="c10"
              >
                Card1
              </div>
            </div>
          </div>
          <div>
            Some child
          </div>
          <div>
            Another child
          </div>
        </div>
        <div
          className="c4"
        >
          <div
            className="c5"
          >
            <div
              className="c6"
            >
              <div
                className="c7"
              />
            </div>
            <div
              className="c8"
            >
              <div
                className="c9"
              >
                Card2
              </div>
              <div
                className="c10"
              >
                Card2
              </div>
            </div>
          </div>
          <div>
            Some child
          </div>
          <div>
            Another child
          </div>
        </div>
        <div
          className="c4"
        >
          <div
            className="c5"
          >
            <div
              className="c6"
            >
              <div
                className="c7"
              />
            </div>
            <div
              className="c8"
            >
              <div
                className="c9"
              >
                Card3
              </div>
              <div
                className="c10"
              >
                Card3
              </div>
            </div>
          </div>
          <div>
            Some child
          </div>
          <div>
            Another child
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`render carousel message canScrollLeft 1`] = `
.c0.c0.c0 {
  width: 100%;
  position: relative;
  box-sizing: border-box;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: flex-end;
  -webkit-box-align: flex-end;
  -ms-flex-align: flex-end;
  align-items: flex-end;
  -webkit-flex-flow: column nowrap;
  -ms-flex-flow: column nowrap;
  flex-flow: column nowrap;
  -webkit-box-pack: end;
  -webkit-justify-content: flex-end;
  -ms-flex-pack: end;
  justify-content: flex-end;
  padding-right: 10px;
}

.c1.c1.c1 {
  width: 100%;
  overflow: hidden;
}

.c2.c2.c2 {
  height: 100%;
  position: relative;
  padding-bottom: 0px;
  box-sizing: border-box;
  height: calc(100% + 17px);
  margin-bottom: -17px;
}

.c3.c3.c3 {
  overflow-y: hidden;
  overflow-x: scroll;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  padding-bottom: 20px;
}

.c6.c6.c6 {
  width: 100%;
  border-bottom: 0px;
  box-sizing: border-box;
  overflow: hidden;
  position: relative;
}

.c6.c6.c6:after {
  padding-top: 52%;
  display: block;
  content: '';
}

.c7.c7.c7 {
  position: absolute;
  top: 0;
  bottom: 0;
  right: 0;
  left: 0;
  background-repeat: no-repeat;
  background-position: 50% 50%;
  background-size: cover;
  background-image: url(someurl);
}

.c8.c8.c8 {
  width: 100%;
  padding: 8px 6px;
  padding-bottom: 12px;
  box-sizing: border-box;
  height: 92px;
  line-height: normal;
}

.c9.c9.c9 {
  font-size: 14px;
  font-family: Helvetica;
  font-weight: bold;
  overflow: hidden;
  color: black;
}

.c10.c10.c10 {
  font-size: 12px;
  font-family: Helvetica;
  color: white;
}

.c4.c4.c4 {
  margin-left: 10px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  padding-bottom: 8px;
}

.c4.c4.c4:first-child {
  margin-left: 0px;
}

.c4.c4.c4:last-child {
  border-right: 10px solid transparent;
}

.c5.c5.c5 {
  width: 260px;
  border-top-left-radius: 16px;
  border-top-right-radius: 16px;
  border-bottom-left-radius: 16px;
  border-bottom-right-radius: 16px;
  background-color: white;
  box-shadow: 0px 0px 10px -2px rgba(0,0,0,0.3);
  margin: 15px 10px;
  margin-right: 0px;
  overflow: hidden;
}

<div
  className="c0"
  style={
    Object {
      "padding": 0,
      "paddingBottom": 10,
    }
  }
>
  <div
    className="c1"
  >
    <div
      className="c2"
    >
      <div
        className="c3"
      >
        <div
          className="c4"
        >
          <div
            className="c5"
          >
            <div
              className="c6"
            >
              <div
                className="c7"
              />
            </div>
            <div
              className="c8"
            >
              <div
                className="c9"
              >
                Card1
              </div>
              <div
                className="c10"
              >
                Card1
              </div>
            </div>
          </div>
          <div>
            Some child
          </div>
          <div>
            Another child
          </div>
        </div>
        <div
          className="c4"
        >
          <div
            className="c5"
          >
            <div
              className="c6"
            >
              <div
                className="c7"
              />
            </div>
            <div
              className="c8"
            >
              <div
                className="c9"
              >
                Card2
              </div>
              <div
                className="c10"
              >
                Card2
              </div>
            </div>
          </div>
          <div>
            Some child
          </div>
          <div>
            Another child
          </div>
        </div>
        <div
          className="c4"
        >
          <div
            className="c5"
          >
            <div
              className="c6"
            >
              <div
                className="c7"
              />
            </div>
            <div
              className="c8"
            >
              <div
                className="c9"
              >
                Card3
              </div>
              <div
                className="c10"
              >
                Card3
              </div>
            </div>
          </div>
          <div>
            Some child
          </div>
          <div>
            Another child
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
